### PR TITLE
Implement mapMPar

### PR DESCRIPTION
**Pull request by luis3m**
Part of #4886

### CODE REVIEW

**Comment by luis3m** [comment]
Initially tried to implement without `toPull` but I couldn't use `(self >>> producer).runManaged.fork` due to the type parameter constraints runManaged has.

**Comment by luis3m** [comment]
Trying to figure out how to resolve that failing test.

File: streams/shared/src/main/scala/zio/stream/experimental/ZChannel.scala [hunk 1]
```diff
@@ -621,6 +621,59 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
...
+    ZChannel.managed {
+      ZManaged.withChildren { getChildren =>
+        for {
+          _           <- ZManaged.finalizer(getChildren.flatMap(Fiber.interruptAll(_)))
+          queue       <- Queue.bounded[ZIO[Env1, Either[OutErr1, OutDone], OutElem2]](n).toManaged(_.shutdown)
+          errorSignal <- Promise.makeManaged[OutErr1, Nothing]
+          permits     <- Semaphore.make(n.toLong).toManaged_
+          pull        <- self.toPull
+          _ <- pull
+                 .foldM(
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/experimental/ZChannel.scala:636]: A `Die` will make this deadlock. So we should use `foldCauseM` here.

Comment (luis3m) [reply streams/shared/src/main/scala/zio/stream/experimental/ZChannel.scala:636]: Good point, thanks.

**Comment by iravid** [comment]
@luis3m Also I think we could avoid the `chunkN(1)` by doing `self.channel.concatMap(ZChannel.fromChunk(_))` (and implementing `ZChannel.fromChunk` :-))

**Comment by iravid** [comment]
Ah I have `ZChannel.fromChunk` in my branch:
```scala
  def writeAll[Out](outs: Chunk[Out]): ZChannel[Any, Any, Any, Any, Nothing, Out, Unit] = {
    def writer(idx: Int, len: Int): ZChannel[Any, Any, Any, Any, Nothing, Out, Unit] =
      if (idx == len) ZChannel.unit
      else ZChannel.write(outs(idx)) *> writer(idx + 1, len)

    writer(0, outs.size)
  }

```

**Comment by iravid** [comment]
@luis3m could you run `sbt prepare` and look into the failing test?

**Comment by luis3m** [comment]
@iravid all passing now (cc @jdegoes)

**Review by iravid — APPROVED** [review]
Fantastic!!

**Pull request merged**